core/chains/evm/logpoller/log_poller.go

```diff
@@ -0,0 +1,344 @@
+		}
+		utils.RetryWithBackoff(ctx, func() bool {
+			if err := lp.orm.InsertLogs(convertLogs(lp.ec.ChainID(), logs)); err != nil {
+				lp.lggr.Warnw("Unable to insert logs logs, retrying", "err", err, "from", i, "to", min(i+lp.backfillBatchSize, end))
+				return true
+			}
+			return false
+		})
+	}
+	return end + 1
+}
+
+// PollAndSaveLogs On startup/crash current is the first block after the last processed block.
+func (lp *LogPoller) PollAndSaveLogs(ctx context.Context, current int64) int64 {
+	// Get latest block on chain
+	latestBlock, err1 := lp.ec.BlockByNumber(ctx, nil)
+	if err1 != nil {
+		lp.lggr.Warnw("Unable to get latest block", "err", err1, "current", current)
+		return current
+	}
+	latest := latestBlock.Number().Int64()
+	// E.g. 1<-2<-3(current)<-4<-5<-6<-7(latest), finality is 2. So 3,4,5 can be batched.
+	// start = current = 3, end = latest - finality = 7-2 = 5 (inclusive range).
+	if (latest - current) >= lp.finalityDepth {
+		current = lp.backfill(ctx, current, latest-lp.finalityDepth)
+	}
+
+	for current <= latest {
+		block, err2 := lp.ec.BlockByNumber(ctx, big.NewInt(current))
+		if err2 != nil {
+			lp.lggr.Warnw("Unable to get block", "err", err2, "current", current)
+			return current
+		}
+		// Does this block point to the same parent that we have saved?
+		// If not, there was a reorg, so we need to rewind.
+		expectedParent, err2 := lp.orm.SelectBlockByNumber(current - 1)
+		if err2 != nil && !errors.Is(err2, sql.ErrNoRows) {
+			// If err is not a no rows error, assume transient db issue and retry
+			lp.lggr.Warnw("Unable to read latest block saved", "err", err2, "current", current)
+			return current
+		}
+		// We will not have the previous block on initial poll or after a backfill.
+		havePreviousBlock := !errors.Is(err2, sql.ErrNoRows)
+		if havePreviousBlock && !bytes.Equal(block.ParentHash().Bytes(), expectedParent.BlockHash.Bytes()) {
+			// There can be another reorg while we're finding the LCA.
+			// That is ok, since we'll detect it on the next iteration of current.
+			lca, err3 := lp.findLCA(block.ParentHash())
+			if err3 != nil {
+				lp.lggr.Warnw("Unable to find LCA after reorg, retrying", "err", err3)
+				return current
+			}
+
+			// We truncate all the blocks and logs after the LCA.
+			// We could preserve the logs for forensics, since its possible
+			// that applications see them and take action upon it, however that
+			// results in significantly slower reads since we must then compute
+			// the canonical set per read. Typically if an application took action on a log
+			// it would be saved elsewhere e.g. eth_txes, so it seems better to just support the fast reads.
+			// Its also nicely analogous to reading from the chain itself.
+			err3 = lp.orm.q.Transaction(func(tx pg.Queryer) error {
+				// These deletes are bounded by reorg depth, so they are
+				// fast and should not slow down the log readers.
+				err3 = lp.orm.DeleteRangeBlocks(lca+1, latest, pg.WithQueryer(tx))
```

Comment:
Let's say there's a backfill and then polling continues as usual.

What guarantee do we have that when this delete happens, the consuming application has already consumed all the backfilled logs it cares about?

Reply:
we only ever backfill finalized blocks, so backfilled logs never get deleted. If an app is querying for logs in unfinalized blocks, they run the risk that those logs could get removed - its on them to record any action taken upon an unfinalized log

Reply:
Ah got it, makes sense